Order Preserving Minimal Perfect Hash Map: Build data structures

KurtMi opened the pull request:
During the build of the OPMPHM a stack and heap is needed. This commit
includes those data structures and there test cases.

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -539,6 +539,54 @@ int elektraKeyLock (Key * key, enum elektraLockOptions what);
+int elektraVstackPush (Vstack * stack, void * data);
+void * elektraVstackPop (Vstack * stack);
+int elektraVstackIsEmpty (Vstack * stack);
+void elektraVstackDestroy (Vstack * stack);
```
Destroy would be new terminology, please use Del or Close.

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -539,6 +539,54 @@ int elektraKeyLock (Key * key, enum elektraLockOptions what);
+void elektraVstackDestroy (Vstack * stack);
+
+/**
+ * The Vheap structure. Used in the OPMPHM.
```
Please explain what a VHeap is, its not so important where it is used (should be in detail and not in brief)

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -539,6 +539,54 @@ int elektraKeyLock (Key * key, enum elektraLockOptions what);
 
 ssize_t ksSearchInternal (const KeySet * ks, const Key * toAppend);
 
+/**
```
whats missing here is the discussed `#ifdef` to avoid Elektra getting larger for embedded systems. Maybe even go for an extra header file?

KurtMi replied:
I am aware of the embedded systems binary size, but I thought even they could also use this code.
I will move it to another header file.

markus2330 replied:
I did not check what the actual size change is, but as long you are the only user such a macro will not hurt.

markus2330 commented on src/libs/elektra/vheap.c:
```diff
@@ -0,0 +1,138 @@
+#include "kdbprivate.h"
+
+/**
+ * Inits the Vheap.
```
Also brief should contain a little bit more information than the function name already has.

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -539,6 +539,54 @@ int elektraKeyLock (Key * key, enum elektraLockOptions what);
+	int minSize;		      /*!< the minimal size of the heap > */
+	int size;		      /*!< The size of the heap */
+	int count;		      /*!< The current number of elements in the heap */
+	int (*comp) (void *, void *); /*!< The comparison function */
```
should be a typedef, it is used multiple times.

markus2330 commented:
Nice, the first part is finally done! I only looked brief at the header file. More comments from me later.

KurtMi commented:
A separate head file will follow, but not in this PR. The CMakeList of the includes looks very difficult.

markus2330 commented on cmake/ElektraCache.cmake:
```diff
@@ -277,6 +277,14 @@ option (INSTALL_SYSTEM_FILES "Install files to system directories" ON)
 
 option (INSTALL_BUILD_TOOLS "Install build tools for cross-compilation" OFF)
 
+option (HASHLOOKUP "Add the possibility to lookup with an Order Preserving Minimal Perfect Hash Map" ON)
```
I would prefer a more generic option and a name which is more consistent with other options.

What about `ENABLE_OPTIMIZATIONS` with the comment "Tturn on optimizations that trade memory for speed"?

markus2330 commented on cmake/ElektraCache.cmake:
```diff
@@ -277,6 +277,14 @@ option (INSTALL_SYSTEM_FILES "Install files to system directories" ON)
 
+option (HASHLOOKUP "Add the possibility to lookup with an Order Preserving Minimal Perfect Hash Map" ON)
+if (HASHLOOKUP)
+	add_definitions(-DHASHLOOKUP=1)
```
Macros should have an `ELEKTRA_` prefix. Furthermore it would be better if you only set it for the source files where needed or define it via kdbconfig.h

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -539,6 +539,63 @@ int elektraKeyLock (Key * key, enum elektraLockOptions what);
+{
+	int minSize;    /*!< the minimal size of the heap > */
+	int size;       /*!< The size of the heap */
+	int count;      /*!< The current number of elements in the heap */
```
should be size_t or ssize_t.

markus2330 commented on src/libs/elektra/CMakeLists.txt:
```diff
@@ -50,6 +50,12 @@ file (GLOB SRC_FILES *.c)
+#remove the opmphm files
+if (NOT HASHLOOKUP)
+	file (GLOB OPMPHM_FILES opmphm*.c)
+	list (REMOVE_ITEM SRC_FILES ${OPMPHM_FILES} ${WIN32_OPMPHM_FILES})
```
What are WIN32_OPMPHM_FILES?

KurtMi replied:
Not sure, I copied from the previous part.

markus2330 replied:
The previous part had exceptions for WIN32, but I do not think you need it ;)

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+#include "kdbprivate.h"
+
+/**
+ * Allocates all space needed.
```
Very vague description. Do you mean "Allocates vheap with size as specified by parameter?"

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+			return 0;
+		}
+	}
+	unsigned int parent, pos;
```
as above, should be size_t or ssize_t.

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+	unsigned int parent, pos;
+	for (pos = vheap->count - 1; pos; pos = parent)
+	{
+		parent = (pos - 1) >> 1;
```
please comment such expressions

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+	if (elektraVheapIsEmpty (vheap)) return NULL;
+	// shrink
+	--vheap->count;
+	if (vheap->size > vheap->minSize && vheap->count <= vheap->size >> 2)
```
again, what are you testing here? (comment missing)

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+		}
+		else
+		{
+			break;
```
you can easily avoid the break by storing the result of the expression in a variable.

KurtMi replied:
not sure what you mean

markus2330 replied:
I mean storing the result of `vheap->comp (vheap->data[child], vheap->data[elem]` in a variable and use it both for the if and to terminate the loop. Then the `break` is not needed anymore.

markus2330 commented on tests/ctest/test_opmphm_vheap.c:
```diff
@@ -0,0 +1,272 @@
+
+#include <tests_internal.h>
+
+int maxcomp (void * a, void * b)
```
why not static? A very short description would not hurt, even if it is only for unit test.

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+ * @return 1 on empty
+ * @return 0 on non empty
+ */
+int elektraVheapIsEmpty (Vheap * vheap)
```
missing const

markus2330 commented:
Looks very nice now! Can you also add some words about the new data structures in `doc/help/elektra-data-structures.md`?

markus2330 commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,80 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+The allocation doubles the space if the `Vstack` is full and reduces the space by the half
+if the memory usage gets under a quarter.
+
+The data will be stored as void pointer and can be inserted with the push call.
```
A small tip: If you use the word invocation instead of call it cannot be confused with the verb call.
(Not necessary to be changed here: it is clear as it is)

markus2330 commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,80 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+		size_t size;
+		void ** data;
+		void ** head;
+	} Vstack;
```
I am not sure if this duplication is a good idea. Did you try to use tracability links to the header file and source files instead?

KurtMi replied:
The split data structure does the same.

markus2330 commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,80 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+
+- **data**: Is the array in which the data will be stored.
+
+- **head**: Is the stack pointer for fast insertion and retrieve. After each push or pop the stack pointer gets incremented or decremented.
```
Again: maybe this additional docu is better placed in the doxygen comments directly at the struct.

markus2330 commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,80 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+At the removal the fist element in the data array will be taken and the data gets reordered.
+The ordering takes log (n) time so the complexity for the insert and remove is log (n).
+
+Beside the classical init and del functions a is empty check is also implemented.
```
Is this part of the example? Maybe use a subsection for the example.

markus2330 commented on doc/help/elektra-data-structures.md:
```diff
@@ -377,26 +377,7 @@ if the memory usage gets under a quarter.
-- **data**: Is the array in which the data will be stored.
-
-- **head**: Is the stack pointer for fast insertion and retrieve. After each push or pop the stack pointer gets incremented or decremented.
+Besides the classical init and del functions, an is-empty check is also implemented.
```
Please add some tracability links to your source + header.

KurtMi replied:
I am not sure what a tracability link is. Is it a note? Kind of `See: src/libs/elektra/opmphm_vstack.c` ?

markus2330 replied:
Yes, please list all relevant source material that has to do with the docu. Going in both directions might be useful, too. Would be great if doxygen can make actual links out if it.

You can think of this docu as requirements and creating links to other artefacts as described in https://en.wikipedia.org/wiki/Requirements_traceability (I suppose you learned that in some requirements lecture already.)

KurtMi replied:
OK, i added references in both directions, but doxygen does not support links to md files.

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -6,7 +6,7 @@
  * To construct a max heap the comparison function VheapComp (a, b), must return
  * 1 on a > b and 0 otherwise. For a min heap 1 on a < b and 0 otherwise.
  *
- * @ingroup vheap
```
Maybe we can add a group "Internal Datastructures" and vheap, vstack and trie are subgroups?

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+
+## Order Preserving Minimal Perfect Hash Map
+
+All structs are defined [in](/src/include/opmphm.h).
```
The link here will look something like this:

> All structs are defined [in](/src/include/opmphm.h).

I think it would be better if you used the following Markdown code:

```Markdown
All structs are defined in [opmphm.h](/src/include/opmphm.h).
```

This way, the sentence is correct even after we convert the Markdown code:

> All structs are defined in [opmphm.h](/src/include/opmphm.h).

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+### Vstack
+
+The `Vstack` is a virtual stack implementation with dynamical memory allocation.
+The allocation doubles the space if the `Vstack` is full and reduces the space by the half
```
> … and reduces the space by **the** half if the memory usage **gets under** a quarter.

Maybe better: and reduces the space by half if the memory usage **drops below** a quarter.

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+Besides the classical init and del functions, an is-empty check is also implemented.
+
+The following [link](/src/libs/elektra/opmphm_vstack.c) leads to the code and
+[this](@ref Vstack) one to the docu.
```
> …**one** to the docu.

Typo: **on** to the docu.

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+
+### Vheap
+
+The `Vheap` is a virtual heap implementation with dynamical memory allocation.
```
> **dynamical** memory allocation.

Maybe better: **dynamic** memory allocation.

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+The allocation works like the `Vstack` allocation.
+
+The data will be stored as void pointer and can be inserted with the insert call.
+At the insert the data will be stored in a binary tree, where each parent and his childs obey the order.
```
> At the **insert** the data will be stored in a binary tree, where each parent and his **childs** obey the order.

Maybe better: At **the point of insertion** the data will be stored in a binary tree, where each parent and his **children** obey the order.

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+At the removal the fist element in the data array will be taken and the data gets reordered.
+The ordering takes log (n) time so the complexity for the insert and remove is log (n).
+
+Beside the classical init and del functions a is-empty check is also implemented.
```
> **a** is-empty check

Typo: **an** is-empty check

sanssecours commented on doc/help/elektra-data-structures.md:
```diff
@@ -366,3 +366,46 @@ The algorithm uses the `syncbits` to decide if the key set needs to be
+[this](@ref Vheap) one to the docu.
+
+####Example
+If the order function constructs a maximum heap it holds for all elements if they have any childs:
```
> If the order function constructs a maximum heap it holds for all elements if they have any **childs**:

Maybe better: If the order function constructs a maximum heap, then the following holds for all elements that have **children**:

sanssecours commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+int elektraVheapIsEmpty (const Vheap * vheap)
+{
+	if (!vheap) return 0;
+	return (vheap->count == 0) ? 1 : 0;
```
I think it would be even easier if we just use `!vheap->count` instead of `(vheap->count == 0) ? 1 : 0` here.

Overall I would replace the two lines with something like this:

```C
// A heap is empty if it exists and its size is 0
return vheap && vheap->count == 0;
```

In theory it should just work as well. At least `test_opmphm_vheap` still works after I change the code. Please keep in mind that this is just a suggestion :o). If you like the code as it is now, there is no need to change it.

sanssecours commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+	int comp = 1;
+	while (comp && child < vheap->count)
+	{
+		if (child + 1 < vheap->count)
```
I think this looks even better if you combine the two if-statements:

```c
if (child + 1 < vheap->count && vheap->comp (vheap->data[child + 1], vheap->data[child]))
{
	++child;
}
```

sanssecours commented on tests/ctest/test_opmphm_vheap.c:
```diff
@@ -0,0 +1,278 @@
+static int maxcomp (void * a, void * b)
+{
+	if (!a || !b) return 1;
+	if (*((int *)a) > *((int *)b))
```
Maybe simpler: Just use

```c
return (*((int *)a) > *((int *)b))
```

instead of second if-statement.

sanssecours commented on tests/ctest/test_opmphm_vheap.c:
```diff
@@ -0,0 +1,278 @@
+static int mincomp (void * a, void * b)
+{
+	if (!a || !b) return 1;
+	if (*((int *)a) < *((int *)b))
```
Maybe simpler: Just use

```c
return (*((int *)a) < *((int *)b))
```

instead of second if-statement.

sanssecours commented on tests/ctest/test_opmphm_vheap.c:
```diff
@@ -0,0 +1,278 @@
+			if (i > actualSize)
+			{
+				// grow
+				actualSize <<= 1;
```
I think you can just use `*=2` here. There is no need to hand optimize a test, as far as I know.

sanssecours commented:
As far as I can tell everything looks fine. I really like that you added a detailed test-suite for your code. The Doxygen comments and the overall documentation are also very nice. 

Some things I noticed overall:
 
- Many commit messages use the same headline. I think it would be better if you use different headings, like in your last two commits. If you want you can also squash (`fixup`) the commits that contain the same headline. You can use something like 
  
  ```sh
  # Please replace 42 with the number of recent commits you want to change
  git rebase --interactive HEAD~42
  ```

  to do that.
  
- There is a typo int the commit message of `90c081b`:

  > This commit includes those data structures and **there** test cases.
  
  Possible Fix: This commit includes those data structures and test cases.

waht commented on tests/ctest/test_opmphm_vstack.c:
```diff
@@ -0,0 +1,134 @@
+static void test_empty ()
+{
+	Vstack * s = elektraVstackInit (5);
+	exit_if_fail (s, "vstack ini error");
```
typo: ini**t**

waht commented on tests/ctest/test_opmphm_vstack.c:
```diff
@@ -0,0 +1,134 @@
+static void test_data ()
+{
+	Vstack * s = elektraVstackInit (10);
+	exit_if_fail (s, "vstack ini error");
```
typo: ini**t**

waht commented on tests/ctest/test_opmphm_vstack.c:
```diff
@@ -0,0 +1,134 @@
+	{
+		int size = minSize;
+		Vstack * s = elektraVstackInit (minSize);
+		exit_if_fail (s, "vstack ini error");
```
typo: ini**t**

waht commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+		vheap->size <<= 1;
+		if (elektraRealloc ((void **)&vheap->data, vheap->size * sizeof (void *)) == -1)
+		{
+			return 0;
```
I'm not sure if this is really an issue, but in line 71 the heap count is incremented and not decremented (i.e. rollback) here before returning

markus2330 commented on src/libs/elektra/opmphm_vheap.c:
```diff
@@ -0,0 +1,141 @@
+ * Checks if the heap is empty.
+ *
+ * @ingroup vheap
+ * @return 1 on empty
```
retval

KurtMi commented:
@markus2330: I added the feedback and a small feature, please have a look at the last commit and merge if nothing more is to change.

markus2330 commented:
Looks fine. A small thing: `opmphm.h` should be renamed to have a prefix, I would suggest `kdbopmphm.h`. I can merge the PR nevertheless, but It should be fixed before we start to install the header file.

KurtMi commented:
> Looks fine. A small thing: opmphm.h should be renamed to have a prefix, I would suggest kdbopmphm.h. I can merge the PR nevertheless, but It should be fixed before we start to install the header file.

I renamed it and squashed all commits to one.

markus2330 commented:
Mmh, we discussed this lengthly. The commits are valueable to understand the discussions, initial intentions and typically problems that occur. Is there any way to restore the commits? (Do you have a backup of the git repo or not yet runned `git gc`)

KurtMi commented:
I made a backup before and restored it!

markus2330 commented:
Thank you very nice contribution! Well done! We hope to see opmphm in action soon ;)

KurtMi commented:
As you maybe know the build process contains 3 sub processes:

* mapping
* ordering
* searching

The first two a more ore less clear and will follow soon as PR (for each one a pr). The last step is a little bit tricky due to the classical inconsistency in the papers, but there is time to figure something out.

markus2330 commented:
Sounds good! It might be useful to move these 3 points to a meta-issue where you keep track of the overall implementation progress, e.g. like Peter does in #475.

KurtMi commented:
Nice Idea, I will start such a ticket, sooner or later.
